applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_solver.py

```diff
@@ -0,0 +1,220 @@
+            self.main_model_part = KratosMultiphysics.ModelPart(model_part_name)
+
+        domain_size = self.settings["domain_size"].GetInt()
+        self.main_model_part.ProcessInfo.SetValue(KratosMultiphysics.DOMAIN_SIZE, domain_size)
+
+    def AddVariables(self):
+        raise Exception("Trying to call AdjointFluidSolver.AddVariables(). Implement the AddVariables() method in the specific derived solver.")
+
+    def AddDofs(self):
+        KratosMultiphysics.VariableUtils().AddDof(KratosMultiphysics.ADJOINT_VELOCITY_X, self.main_model_part)
+        KratosMultiphysics.VariableUtils().AddDof(KratosMultiphysics.ADJOINT_VELOCITY_Y, self.main_model_part)
+        KratosMultiphysics.VariableUtils().AddDof(KratosMultiphysics.ADJOINT_VELOCITY_Z, self.main_model_part)
+        KratosMultiphysics.VariableUtils().AddDof(KratosMultiphysics.ADJOINT_PRESSURE, self.main_model_part)
+
+        if self._IsPrintingRank():
+            KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Adjoint fluid solver DOFs added correctly.")
+
+    def ImportModelPart(self):
+        # we can use the default implementation in the base class
+        self._ImportModelPart(self.main_model_part,self.settings["model_import_settings"])
+
+    def PrepareModelPart(self):
+        if not self.main_model_part.ProcessInfo[KratosMultiphysics.IS_RESTARTED]:
+            ## Replace default elements and conditions
+            self._ReplaceElementsAndConditions()
+            ## Executes the check and prepare model process
+            self._ExecuteCheckAndPrepare()
+            ## Set buffer size
+            self.main_model_part.SetBufferSize(self.min_buffer_size)
+
+        if not self.model.HasModelPart(self.settings["model_part_name"].GetString()):
+            self.model.AddModelPart(self.main_model_part)
+
+        if self._IsPrintingRank():
+            KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Model reading finished.")
+
+    def ExportModelPart(self):
+        ## Model part writing
+        name_out_file = self.settings["model_import_settings"]["input_filename"].GetString()+".out"
+        KratosMultiphysics.ModelPartIO(name_out_file, KratosMultiphysics.IO.WRITE).WriteModelPart(self.main_model_part)
+
+        if self._IsPrintingRank():
+            KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Model export finished.")
+
+    def GetMinimumBufferSize(self):
+        return self.min_buffer_size
+
+    def Initialize(self):
+        raise Exception("Calling AdjointFluidSolver.Initialize() base method. Please implement a custom Initialize() method for your solver.")
+
+    def AdvanceInTime(self, current_time):
+        dt = self._ComputeDeltaTime()
+        new_time = current_time + dt
+
+        self.main_model_part.CloneTimeStep(new_time)
+        self.main_model_part.ProcessInfo[KratosMultiphysics.STEP] += 1
+
+        return new_time
+
+    def InitializeSolutionStep(self):
+        self.solver.InitializeSolutionStep()
```

Comment:
I would highly recommend changing this name!

self.solver is a solving strategy, it becomes really confusing since the class itself is also a solver

=> you can take a look at the solvers in StructuralMechanics